tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -1114,6 +1343,69 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
     }
   }
 
+  // Check if the optimistic analysis converges.  Specifically, check whether
+  // all the predicates of the merge nodes in the same frame are the same.  If
+  // yes, report success.  If not, report failure and clear the assigned
+  // predicates.
+  if (use_optimistic_mode) {
```

Comment:
I think this will be slightly more robust / obviously correct if in `HandleMerge` we keep track of the nodes we assigned optimistic predicates to (in some `std::vector<Node*>` field in the class) and check that those specific nodes converged to the same predicate here.

Also, maybe we can move this to a helper function?

Reply:
I'd argue that the robustness is the same but the current implementation is more clear, because we explicitly check all the merges are the same for each frame. Note that in the case of nested while loops, each while can have different predicates while merges on the same level of while must have the same predicate (I added some comments to make this more explicit.) This can complicate the track and compare approach.